pkg/sql/sqlbase/testutils.go

```diff
@@ -375,3 +379,169 @@ func TestingMakePrimaryIndexKey(desc *TableDescriptor, vals ...interface{}) (roa
 func TestingDatumTypeToColumnSemanticType(ptyp types.T) (ColumnType_SemanticType, error) {
 	return datumTypeToColumnSemanticType(ptyp)
 }
+
+// RandCreateTable creates a random CreateTable definition.
+func RandCreateTable(tableIdx int) *tree.CreateTable {
+	rand := rand.New(rand.NewSource(timeutil.Now().UnixNano()))
+	// columnDefs contains the list of Columns we'll add to our table.
+	columnDefs := make([]*tree.ColumnTableDef, randutil.RandIntInRange(rand, 1, 20))
+	// defs contains the list of Columns and other attributes (indexes, column
+	// families, etc) we'll add to our table.
+	defs := make(tree.TableDefs, len(columnDefs))
+
+	for i := range columnDefs {
+		columnDef := randColumnTableDef(rand, i)
+		columnDefs[i] = columnDef
+		defs[i] = columnDef
+	}
+
+	// Shuffle our column definitions in preparation for random partitioning into
+	// column families.
+	rand.Shuffle(len(columnDefs), func(i, j int) {
+		columnDefs[i], columnDefs[j] = columnDefs[j], columnDefs[i]
+	})
+
+	// Partition into column families.
+	numColFams := randNumColFams(rand, len(columnDefs))
+
+	// Create a slice of indexes into the columnDefs slice. We'll use this to make
+	// the random partitioning by picking some indexes at random to use as
+	// partitions boundaries.
+	indexes := make([]int, len(columnDefs)-1)
+	for i := range indexes {
+		indexes[i] = i + 1
+	}
+	rand.Shuffle(len(indexes), func(i, j int) {
+		indexes[i], indexes[j] = indexes[j], indexes[i]
+	})
+
+	// Grab our random partition boundaries, and re-sort back into sorted index
+	// order.
+	numSeparators := numColFams - 1
+	indexes = indexes[:numSeparators]
+	sort.Slice(indexes, func(i, j int) bool {
+		return indexes[i] < indexes[j]
+	})
+
+	indexesWithZero := make([]int, len(indexes)+2)
+	copy(indexesWithZero[1:], indexes)
+	indexesWithZero[len(indexesWithZero)-1] = len(columnDefs)
+	indexes = indexesWithZero
+
+	// Now (finally), indexes is the list of partitions we're going to slice the
+	// column def list into. Create our column families by grabbing the slice of
+	// columns from the column list bounded by each partition index at the end.
+	// Also, save column family 0 for later as all primary keys have to be part of
+	// that column family.
+	var colFamZero []*tree.ColumnTableDef
+	for i := 0; i+1 < len(indexes); i++ {
+		start, end := indexes[i], indexes[i+1]
+
+		names := make(tree.NameList, end-start)
+		for j := start; j < end; j++ {
+			names[j-start] = columnDefs[j].Name
+		}
+		if colFamZero == nil {
+			for j := start; j < end; j++ {
+				colFamZero = append(colFamZero, columnDefs[j])
+			}
+		}
+
+		famDef := &tree.FamilyTableDef{
+			Name:    tree.Name(fmt.Sprintf("fam%d", i)),
+			Columns: names,
+		}
+		defs = append(defs, famDef)
+	}
+
+	// Make a random primary key with high likelihood.
+	if rand.Intn(8) != 0 {
+		indexDef := randIndexTableDefFromCols(rand, colFamZero)
+		if len(indexDef.Columns) > 0 {
+			defs = append(defs, &tree.UniqueConstraintTableDef{
```

Comment:
won't this sometimes make multiple primary keys? how does that work?

Reply:
What do you mean? This isn't in a loop or anything - as far as I can tell, no, it'll either make 0 or 1 primary key. Maybe you're talking about how the pk can have multiple columns?

Reply:
I misread and thought this was in the for loop.